api/server.go

```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
+		return
+	}
+
+	var siacoinInput types.Currency
+	for i, sco := range wcr.Siacoins {
+		switch {
+		case sco.Value.IsZero():
+			jc.Error(fmt.Errorf("siacoin output %d has zero value", i), http.StatusBadRequest)
+			return
+		case sco.Address == types.VoidAddress:
+			jc.Error(fmt.Errorf("siacoin output %d has void address", i), http.StatusBadRequest)
+			return
+		}
+		siacoinInput = siacoinInput.Add(sco.Value)
+	}
+
+	var siafundInput uint64
+	for i, sfo := range wcr.Siafunds {
+		switch {
+		case sfo.Value == 0:
+			jc.Error(fmt.Errorf("siafund output %d has zero value", i), http.StatusBadRequest)
+			return
+		case sfo.Address == types.VoidAddress:
+			jc.Error(fmt.Errorf("siafund output %d has void address", i), http.StatusBadRequest)
+			return
+		}
+		siafundInput += sfo.Value
+	}
+
+	if siacoinInput.IsZero() && siafundInput == 0 {
+		jc.Error(errors.New("no inputs provided"), http.StatusBadRequest)
+	}
+
+	fee := s.cm.RecommendedFee().Mul64(2000) // use a const for simplicity
+
+	var sent bool
+	var locked []types.Hash256
+	defer func() {
+		if sent {
+			return
+		}
+		s.wm.Release(locked)
+	}()
+
+	sces, basis, siacoinChange, err := s.wm.SelectSiacoinElements(walletID, siacoinInput.Add(fee), false)
+	if err != nil {
+		jc.Error(fmt.Errorf("failed to select siacoin elements: %w", err), http.StatusInternalServerError)
+		return
+	}
+	for _, sce := range sces {
+		locked = append(locked, types.Hash256(sce.ID))
+	}
+
+	if !siacoinChange.IsZero() {
+		wcr.Siacoins = append(wcr.Siacoins, types.SiacoinOutput{
+			Value:   siacoinChange,
+			Address: wcr.ChangeAddress,
+		})
+	}
+
+	sfes, _, siafundChange, err := s.wm.SelectSiafundElements(walletID, siafundInput)
+	if err != nil {
+		jc.Error(fmt.Errorf("failed to select siafund elements: %w", err), http.StatusInternalServerError)
+		return
+	}
+	for _, sfe := range sfes {
+		locked = append(locked, types.Hash256(sfe.ID))
+	}
+
+	if siafundChange > 0 {
+		wcr.Siafunds = append(wcr.Siafunds, types.SiafundOutput{
+			Value:   siafundChange,
+			Address: wcr.ChangeAddress,
+		})
+	}
+
+	knownAddresses := make(map[types.Address]wallet.Address)
+	getAddress := func(addr types.Address) (wallet.Address, error) {
+		if a, ok := knownAddresses[addr]; ok {
+			return a, nil
+		}
+		a, err := s.wm.WalletAddress(walletID, addr)
+
+		if err != nil {
+			return wallet.Address{}, err
+		}
+		knownAddresses[addr] = a
+		return a, nil
+	}
+
+	resp := WalletConstructResponse{
+		Basis:        basis,
+		EstimatedFee: fee,
+	}
+
+	txn := types.Transaction{
+		MinerFees:      []types.Currency{fee},
+		SiacoinInputs:  make([]types.SiacoinInput, 0, len(sces)),
+		SiacoinOutputs: wcr.Siacoins,
+		SiafundInputs:  make([]types.SiafundInput, 0, len(sfes)),
+		SiafundOutputs: wcr.Siafunds,
+	}
+
+	for _, sce := range sces {
+		addr, err := getAddress(sce.SiacoinOutput.Address)
+		if err != nil {
+			jc.Error(fmt.Errorf("failed to get address: %w", err), http.StatusInternalServerError)
```

Comment:
want to handle `wallet.ErrNotFound` and return a 404? If we do we might as well make it part of the validation up top